Improve pull-to-refresh usability in RefreshContainer

gentledepp opened the pull request:
<!--- See CONTRIBUTING.md for general guidelines on contributions -->

## What does the pull request do?
<!--- Give a bit of background on the PR here, together with links to with related issues etc. -->

Pull to refresh is non-optimal in its current implementation:
1. you need to select the first item in a list to start a pull2refresh.
2. It is not supported on desktop devices and cannot optionally be added without rewriting a bunch of internal code.

## What is the current behavior?
<!--- If the PR is a fix, describe the current incorrect behavior, otherwise delete this section. -->

 #15529 goes into great detail.
 #17496 is the original pull request - but it seems to be abandoned.
 
 So I: 
 -  cherry-picked all changes from #17496 (the pr seems to be abandoned)
 -  incorporated your feedback (the changes you requested in #17496)
 -  added the possibility, to also enable pull2refresh on desktops (mouse-controlled devices)


## What is the updated/expected behavior with this PR?
<!--- Describe how to test the PR. -->

Pull to refresh works seemlessly for mobile and (optionally) also for desktop devices

## How was the solution implemented (if it's not obvious)?
<!--- Include any information that might be of use to a reviewer here. -->

see #17496 for more details.

## Checklist

- [x] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/avalonia-docs with user documentation

I will consider submitting a PR to the documentation, if the desktop support property is accepted and the name is fixed.

## Breaking changes
<!--- List any breaking changes here. -->

## Obsoletions / Deprecations
<!--- Obsolete and Deprecated attributes on APIs MUST only be included when discussed with Core team. @grokys, @kekekeks & @danwalmsley -->

## Fixed issues
<!--- If the pull request fixes issue(s) list them like this: 
Fixes #123
Fixes #456
-->
Fixes #15529

nickodei committed:
Added improved refresh-capabilities to ScrollViewer in simple theme
Added gesture-recognizer for scroll-viewer to improve usability
Added possibility to enable pulltorefresh on desktop too

cla-avalonia commented:
@alexander.marek,

Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
```
@cla-avalonia agree
```

<details>
    <summary>Contributor License Agreement</summary>
    
### Contribution License Agreement

This Contribution License Agreement ( **“Agreement”** ) is agreed to by the party signing below ( **“You”** ),
and conveys certain license rights to the AvaloniaUI OÜ ( **“AvaloniaUI OÜ”** ) for Your contributions to
AvaloniaUI OÜ open source projects. This Agreement is effective as of the latest signature date below.

**1. Definitions.**

**“Code”** means the computer software code, whether in human-readable or machine-executable form,
that is delivered by You to AvaloniaUI OÜ under this Agreement.

**“Project”** means any of the projects owned or managed by AvaloniaUI OÜ and offered under a license
approved by the Open Source Initiative (www.opensource.org).

**“Submit”** is the act of uploading, submitting, transmitting, or distributing code or other content to any
Project, including but not limited to communication on electronic mailing lists, source code control
systems, and issue tracking systems that are managed by, or on behalf of, the Project for the purpose of
discussing and improving that Project, but excluding communication that is conspicuously marked or
otherwise designated in writing by You as “Not a Submission.”

**“Submission”** means the Code and any other copyrightable material Submitted by You, including any
associated comments and documentation.

**2. Your Submission.** You must agree to the terms of this Agreement before making a Submission to any
Project. This Agreement covers any and all Submissions that You, now or in the future (except as
described in Section 4 below), Submit to any Project.

**3. Originality of Work.**  You represent that each of Your Submissions is entirely Your 
original work. Should You wish to Submit materials that are not Your original work, 
You may Submit them separately to the Project if You (a) retain all copyright and 
license information that was in the materials as you received them, (b) in the 
description accompanying your Submission, include the phrase &quot;Submission 
containing materials of a third party:&quot; followed by the names of the third party and any 
licenses or other restrictions of which You are aware, and (c) follow any other 
instructions in the Project&#39;s written guidelines concerning Submissions.

**4. Your Employer.** References to “employer” in this Agreement include Your employer or anyone else
for whom You are acting in making Your Submission, e.g. as a contractor, vendor, or agent. If Your
Submission is made in the course of Your work for an employer or Your employer has intellectual
property rights in Your Submission by contract or applicable law, You must secure permission from Your
employer to make the Submission before signing this Agreement. In that case, the term “You” in this
Agreement will refer to You and the employer collectively. If You change employers in the future and
desire to Submit additional Submissions for the new employer, then You agree to sign a new Agreement
and secure permission from the new employer before Submitting those Submissions.

**5. Licenses.**

**a. Copyright License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly
or indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable
license in the Submission to reproduce, prepare derivative works of, publicly display, publicly perform,
and distribute the Submission and such derivative works, and to sublicense any or all of the foregoing
rights to third parties.

**b. Patent License.** You grant AvaloniaUI OÜ, and those who receive the Submission directly or
indirectly from AvaloniaUI OÜ, a perpetual, worldwide, non-exclusive, royalty-free, irrevocable license
under Your patent claims that are necessarily infringed by the Submission or the combination of the
Submission with the Project to which it was Submitted to make, have made, use, offer to sell, sell and
import or otherwise dispose of the Submission alone or with the Project.

**c. Other Rights Reserved.** Each party reserves all rights not expressly granted in this Agreement.
No additional licenses or rights whatsoever (including, without limitation, any implied licenses) are
granted by implication, exhaustion, estoppel or otherwise.

**6. Representations and Warranties.** You represent that You are legally entitled to grant the above
licenses. You represent that each of Your Submissions is entirely Your original work (except as You may
have disclosed under Section 3 ). You represent that You have secured permission from Your employer to
make the Submission in cases where Your Submission is made in the course of Your work for Your
employer or Your employer has intellectual property rights in Your Submission by contract or applicable
law. If You are signing this Agreement on behalf of Your employer, You represent and warrant that You
have the necessary authority to bind the listed employer to the obligations contained in this Agreement.
You are not expected to provide support for Your Submission, unless You choose to do so. UNLESS
REQUIRED BY APPLICABLE LAW OR AGREED TO IN WRITING, AND EXCEPT FOR THE WARRANTIES
EXPRESSLY STATED IN SECTIONS 3, 4, AND 6 , THE SUBMISSION PROVIDED UNDER THIS AGREEMENT IS
PROVIDED WITHOUT WARRANTY OF ANY KIND, INCLUDING, BUT NOT LIMITED TO, ANY WARRANTY OF
NONINFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.

**7. Notice to AvaloniaUI OÜ.** You agree to notify AvaloniaUI OÜ in writing of any facts or
circumstances of which You later become aware that would make Your representations in this
Agreement inaccurate in any respect.

**8. Information about Submissions.** You agree that contributions to Projects and information about
contributions may be maintained indefinitely and disclosed publicly, including Your name and other
information that You submit with Your Submission.

**9. Governing Law/Jurisdiction.** This Agreement is governed by the laws of the Republic of Estonia, and
the parties consent to exclusive jurisdiction and venue in the courts sitting in Talinn,
Estonia. The parties waive all defenses of lack of personal jurisdiction and forum non-conveniens.

**10. Entire Agreement/Assignment.** This Agreement is the entire agreement between the parties, and
supersedes any and all prior agreements, understandings or communications, written or oral, between
the parties relating to the subject matter hereof. This Agreement may be assigned by AvaloniaUI OÜ.


*AvaloniaUI OÜ dedicates this Contribution License Agreement to the public domain according to the Creative Commons CC0 1.*

</details>

**1** out of **2** committers have signed the CLA.
- [x] @nickodei
- [ ] @alexander.marek




**alexander.marek** doesn't seem to be a GitHub user.
Please add the emails used in your commits to your Github account accordingly.

thevortexcloud commented on src/Avalonia.Controls/PullToRefresh/ScrollablePullGestureRecognizer.cs:
```diff
@@ -0,0 +1,130 @@
+
+        protected override void PointerPressed(PointerPressedEventArgs e)
+        {
+            if (Target != null && Target is Visual visual) //&& (e.Pointer.Type == PointerType.Touch || e.Pointer.Type == PointerType.Pen))
```
Does this mean it works with a mouse? Or is this just for testing? I personally always found it weird some of the gesture recognisers are touch only.

gentledepp replied:
This is just a draft - please ignore - I needed the nuget package.
Its the pull2refresh PR including support for desktop apps (so yes, pull to refresh works on windows too)

thevortexcloud replied:
Sounds like a significant improvement over the existing pull gesture recogniser then.

gentledepp commented:
> @alexander.marek,
> 
> Please read the following Contributor License Agreement (CLA). If you agree with the CLA, please reply with the following:
> 
> ```
> @cla-avalonia agree
> ```
> 
> Contributor License Agreement
> **1** out of **2** committers have signed the CLA.
> 
> * [x]  @nickodei
> * [ ]  @alexander.marek
> 
> **alexander.marek** doesn't seem to be a GitHub user. Please add the emails used in your commits to your Github account accordingly.

@cla-avalonia agree

gentledepp commented:
This is actually mainly the code of #17496, with one addition:

Added a styled property "IsEnabledOnDesktop" that allows to optionally enable the pulltorefresh gesture on Desktop applications as well.

emmauss commented:
Could you make your pr description more descriptive? Even if it's based on #17496 , it should still follow contribution guidelines.

gentledepp commented:
@emmauss 

Essentially, I 
* cherry-picked all changes from #17496 
* incorporated your feedback (the changes you requested in #17496)
* added the possibility, to also enable pull2refresh on desktops (mouse-controlled devices)

The third feature is important. It allows for
* easier testing of pull2refresh (mac and windows)
* opt in pull2refresh for customers, that are confused why "this works on my phone but not my computer" (yes, they exist...)

IsEnabledOnDesktop is a _styled property_
So it is disabled by default, but can easily be enabled using a global style.

emmauss commented:
> @emmauss
> 
> Essentially, I
> 
>     * cherry-picked all changes from [Improve Refresh-Container usability #17496](https://github.com/AvaloniaUI/Avalonia/pull/17496)
> 
>     * incorporated your feedback (the changes you requested in [Improve Refresh-Container usability #17496](https://github.com/AvaloniaUI/Avalonia/pull/17496))
> 
>     * added the possibility, to also enable pull2refresh on desktops (mouse-controlled devices)
> 
> 
> The third feature is important. It allows for
> 
>     * easier testing of pull2refresh (mac and windows)
> 
>     * opt in pull2refresh for customers, that are confused why "this works on my phone but not my computer" (yes, they exist...)
> 
> 
> IsEnabledOnDesktop is a _styled property_ So it is disabled by default, but can easily be enabled using a global style.
> 
> Note that the propertyname is not of my concern. I understand that you have API design meetings and will most likely change it. You've got my blessings to do so (although you do not need them ;-))

It would be best if you put this in the PR description and follow the section guides.

MrJul commented:
Notes from the API review:

## Remove the properties added to `ScrollGestureRecognizer`

Exposing `Extent`, `Offset` and `Viewport` shouldn't be necessary. We don't want to add every possible scroll info property as part of the public API of the gesture recognizer.

Instead, the `ScrollGestureRecognizer` should listen to property change notifications from its parent presenter, which is accessible through the existing `Target` property.

## Rename `IsEnabledOnDesktop` to `IsMouseEnabled`

The property isn't really related to desktop: it simply enables the mouse to be recognized, even on mobile platforms.

MrJul commented on src/Avalonia.Controls/PullToRefresh/ScrollablePullGestureRecognizer.cs:
```diff
@@ -0,0 +1,136 @@
+        public ScrollablePullGestureRecognizer(PullDirection pullDirection, bool IsMouseEnabled)
+        {
+            PullDirection = pullDirection;
+            IsMouseEnabled = IsMouseEnabled;
```
This assignment is wrong and does nothing.

MrJul commented on src/Avalonia.Controls/PullToRefresh/ScrollablePullGestureRecognizer.cs:
```diff
@@ -0,0 +1,136 @@
+
+        public bool IsMouseEnabled { get; set; }
+
+        public ScrollablePullGestureRecognizer(PullDirection pullDirection, bool IsMouseEnabled)
```
Rename the parameter to `isMouseEnabled`.

emmauss committed:
Merge branch 'master' into feature/15529_pull2refresh

MrJul committed:
Merge branch 'master' into feature/15529_pull2refresh

MrJul committed:
Fix naming

MrJul committed:
Adjust tests to avoid reflection

MrJul reviewed (approved):
Re-tested, LGTM!
